pageserver/src/layered_repository.rs

```diff
@@ -2340,22 +2353,47 @@ impl LayeredTimeline {
         } else if cfg!(test) {
             // We don't have local timeline in mocked cargo tests.
             // So, just ignore pitr_interval setting in this case.
-            pitr_cutoff_lsn = cutoff;
+            pitr_cutoff_lsn = gc_info.horizon_cutoff;
         }
+        gc_info.pitr_cutoff = pitr_cutoff_lsn;
+
```

Comment:
I think you need to also update the timeline's `latest_gc_cutoff_lsn` here. Otherwise essentially the same race condition can still happen:

1. B starts GC and scans the list of timelines. The new timeline doesn't exist yet, so it only sees 'main'
2. B calls update_gc_info() on the source timeline, setting pitr/horizon_cutoff to 1/22220000
3. B releases gc_cs
4. A acquires gc_cs
5. A reads latest_gc_cutoff_lsn as 1/11110000.
6. A finishes creating the new timeline
7. B runs second phase of GC, updates latest_gc_cutoff_lsn to 1/22220000, and removes layers older than that.

Reply:
I don't update the gc cutoff there because it's possible to not GC even after updating GC info. For example, that can happen when we go a pageserver shutdown request: https://github.com/neondatabase/neon/pull/2101/files#diff-2dee9987054422607c1ab7369e102927d8477dea6d75c4c7f3891e3d65b15d1eR971-R975

The situation above should not happen because I also added a check for GC info in `branch_timeline`:
https://github.com/neondatabase/neon/pull/2101/files#diff-2dee9987054422607c1ab7369e102927d8477dea6d75c4c7f3891e3d65b15d1eR306-R312

Also the new test simulates that situation, so it should fail if there is a race condition.